protocol/src/lib.rs

```diff
@@ -227,160 +264,188 @@ impl PacketReader {
         content_slice[0..3].copy_from_slice(&enc_content_len);
         let content_len = u32::from_le_bytes(content_slice);
 
-        // Include 1-byte decoy and 16-byte tag.
-        content_len as usize + NUM_DECOY_BYTES + TAG_BYTES
+        content_len as usize + NUM_HEADER_BYTES + NUM_TAG_BYTES
     }
 
-    /// Decrypt the rest of the message from the peer, excluding the 3 length bytes. This method should only be called after
-    /// calling `decrypt_len` on the first three bytes of the buffer.
+    /// Decrypt the packet contents.
     ///
     /// # Arguments
     ///
-    /// - `ciphertext` - The message from the peer.
-    /// - `contents`   - Mutable buffer to write plaintext.
-    /// - `aad`        - Optional authentication for the peer.
+    /// * `ciphertext` - The packet from the peer excluding the first 3 length bytes. It should contain
+    ///                  the header, contents, and authentication tag.
+    /// * `contents`   - Mutable buffer to write plaintext. Note that the first byte is the header byte
+    ///                  containing protocol flags.
+    /// * `aad`        - Optional associated authenticated data.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing:
+    ///   * `Ok(PacketType)`: A flag indicating if the decoded packet is a decoy or not.
+    ///   * `Err(Error)`: An error that occurred during decryption.
     ///
     /// # Errors
     ///
-    /// Fails if the packet was not decrypted or authenticated properly.  
+    /// * `CiphertextTooSmall` - Ciphertext argument does not contain a whole packet.
+    /// * `BufferTooSmall `    - Contents buffer argument is not large enough for plaintext.
+    /// * Decryption errors for any failures such as a tag mismatch.
     pub fn decrypt_contents(
         &mut self,
         ciphertext: &[u8],
         contents: &mut [u8],
         aad: Option<&[u8]>,
-    ) -> Result<(), Error> {
+    ) -> Result<PacketType, Error> {
```

Comment:
I wish we were able to take full advantage of the type system here and have `PacketType::Genuine` actually return the buffer within the variant, but we have an unwavering support for `nostd` lol.

Reply:
We could probably do that if we took ownership of the `contents` buffer here, but I think that might create confusion on the caller side? Could no longer pass in a reference to a socket or whatever.

Reply:
All good, I think most will just use the `with_alloc` variants. Since I have only ever seen projects that use the `std` library that could even make use of this, I may push for a renaming from `with_alloc` functions to be named normally and for `nostd` functions to have the suffix `no_alloc`.

Reply:
Yep, that makes sense to me.